### DISCUSSION TITLE
Fundamentals of SK - How does xxxxx work? A short list of questions.

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#611] [Q&A]
Been working with SK for some time now and have simply used the Demo as a basis application, not really understanding much of the inner workings. I would like to ask dumb questions about thing that most will think obvious.
How do the Initialize, Update, Shutdown methods get called. I have an Interface IRun (just like SK's Demo ITest). I derive my class from IRun and implement a simple cube in the Update method. What is the mechanism that calls Update? I know what Step does.

In the Tools folder of the SK project there are several classes. HolographicTheme.cs has no references? What is it for?
RenderCamera.cs, looking at the code it seems to be taking snapshot of frames? Is that what it does? What is its purpose?
I tried the ThemeEditor - cool. You can colorize most everything but button faces? I couldn't find a way to make fixed size buttons?

I want to work with Eye (gaze) to select an object similar to the Unity version and Microsoft Guides, how can you achieve some of those kinds of things in SK?

Is there a good place to go see what others have done with SK, I've been on Discord but that forum doesn't seem the place?

**Comment by maluoi** [comment]
> How do the Initialize, Update, Shutdown methods get called. I have an Interface IRun (just like SK's Demo ITest). I derive my class from IRun and implement a simple cube in the Update method. What is the mechanism that calls Update?

The only event you normally get from StereoKit is the main Step loop! Every single frame this gets called, and everything else is just derived from that. The lifecycle of StereoKitTest's `ITest` objects is managed by the static [`Tests`](https://github.com/StereoKit/StereoKit/blob/master/Examples/StereoKitTest/Tests.cs) class, which itself is managed by a quite simple pattern:

```csharp
SK.Initialize(...);

Tests.Initialize();

SK.Run(()=>{
    Tests.Step()
});

Tests.Shutdown();
```

There's also a small project for the [v0.3.1 release notes](https://github.com/StereoKit/StereoKitReleaseNotes/tree/main/v0.3.1) that uses an `IScene`, which is basically the same as the `ITest`, but simplified a little to remove some of the testing functionality. That was simple enough that I didn't even make a class for it, all the lifecycle code there was in [Program.cs](https://github.com/StereoKit/StereoKitReleaseNotes/blob/main/v0.3.1/Program.cs).

The `IStepper` is exclusive to the C# API of StereoKit, and I have this around largely so that it's easy to build and provide tools with a single attachment point! Most of the code in the `Tools` folder implement this, and can be added to your project and forgotten about with a simple `SK.AddStepper<RenderCamera>()` sort of line. Technically, it's _very_ similar to the `IScene` or `ITest` sample code, but it's part of SK's core, so you can't extend it or add additional control over its lifecycle. It also has a lot more considerations for "what if a developer did this strange thing?", like for example, adding an `IStepper` from a separate thread before initialization. So it's a pretty robust system with a bit more weight and complexity. Recent preview versions of the [`IStepper` have commentdocs](https://github.com/StereoKit/StereoKit/blob/develop/StereoKit/Framework/IStepper.cs) that go into some additional detail about how their lifetimes work.

`IStepper` is a nice thing to use, but I like to think of an `IStepper` as a system, rather than an object? You could treat it like an object if you wanted to, but I suspect you'd find it somewhat limited in functionality there. It's also somewhat object oriented, so it comes with the performance overhead of using classes by reference and virtual methods. Much like a Unity GameObject would.

---

> [HolographicTheme.cs](https://github.com/StereoKit/StereoKit/blob/develop/Examples/StereoKitTest/Tools/HolographicTheme.cs) has no references? What is it for?

This is the MRDL-like theme for StereoKit! I believe you're familiar with [Finn's fork](https://github.com/Zee2/StereoKitMRDL) of it. It is not directly used in StereoKitTest, due to how the `Tests` screenshot generations works for documentation, but I'll turn it on now and again in development to make sure the more robust theming features still work well. I've also just generally been keeping single-file tool code in that folder for other people's benefit.

---

> RenderCamera.cs, looking at the code it seems to be taking snapshot of frames? Is that what it does? What is its purpose?

I use this one for making .gifs! You can place the camera in your scene, point it, change the field of view, and record a stream of bitmaps. It works super nicely for getting the third person perspective, and in particular gifs like this tend to compress really nicely! :)

This will basically just render everything in your scene to a render texture, from a vantage point of your choosing. Then it'll pull the color data from the texture and write it to file. Right now it's just bitmaps, I think a more robust camera would probably use a library to write directly to a gif or video file. The current bitmap solution can be a bit laggy. I guess it's also good to not that this does _not_ composite with the real world if you're using AR, since it can render from any location regardless of where the actual headset is located.

---

> I tried the ThemeEditor - cool. You can colorize most everything but button faces? I couldn't find a way to make fixed size buttons?

You should be able to do button faces! Not sure what you mean about fixed size buttons though, you can do this by passing a size to the buttons, but it's unrelated to theming.

---

> I want to work with Eye (gaze) to select an object similar to the Unity version and Microsoft Guides, how can you achieve some of those kinds of things in SK?

There's [Input.Eyes](https://stereokit.net/Pages/StereoKit/Input/Eyes.html), but this isn't tied into the UI yet. Later it will be :)

---

> Is there a good place to go see what others have done with SK, I've been on Discord but that forum doesn't seem the place?

Discord does have a showcase channel now! You might also want to check out GitHub's [Used By](https://github.com/StereoKit/StereoKit/network/dependents) list.
